src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -189,12 +194,17 @@ internal abstract partial class AnalyzerDriver : IDisposable
         /// </summary>
         /// <param name="analyzers">The set of analyzers to include in the analysis</param>
         /// <param name="analyzerManager">AnalyzerManager to manage analyzers for analyzer host's lifetime.</param>
+        /// <param name="filteredSeverities">Filtered diagnostic severities in the compilation, i.e. diagnostics with effective severity from this set should not be reported.</param>
         /// <param name="isComment">Delegate to identify if the given trivia is a comment.</param>
-        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerManager analyzerManager, Func<SyntaxTrivia, bool> isComment)
+        protected AnalyzerDriver(ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerManager analyzerManager, ImmutableHashSet<ReportDiagnostic> filteredSeverities, Func<SyntaxTrivia, bool> isComment)
         {
+            Debug.Assert(!filteredSeverities.Contains(ReportDiagnostic.Suppress));
+            Debug.Assert(!filteredSeverities.Contains(ReportDiagnostic.Default));
```

Comment:
Reading the PR i'm wondering if we are being overly general here. The API here takes a data type that implies that it can filter out any severity but internally limits it to effectively only two severities. In the code base we only ever provide two combinations of this value. Using a hash table style structure here seems potentially overkill for a dataset that has a max size of two. 

Should we consider instead just making this a simple flags `enum` or two `bool` that filter out the two diagnostic severities that we care about? Or am I missing a scenario here which dictates we should be using the `ImmutableHashSet<T>`?

Reply:
The primary reason to choose a hash set was for simplicity of implementation in this method - we can just do "filteredSeverity.Contains(severity)" check everywhere without worrying about each specific severity. Contains check should be pretty efficient as well.

With 2 bools approach, we need to map each bool to a corresponding severity and have checks such as `if (filterHiddenSeverity && severity == ReportDiagnostic.Hidden || filterInfoSeverity && severity == ReportDiagnostic.Info)`. Similarly, with enum approach, we likely need a flags enum with bit more complex code. I can certainly switch to one of these approaches if you feel that would be more readable and efficient then the current approach.

Reply:
Hash table still seems heavy here. Did you consider a simple `struct` or `enum` flag that could do the same thing? For instance something like:

```cs
readonly struct HiddenFilter
{
  private readonly int _flag;
  private HiddenFilter(bool includeHidden, bool includeInfo) {
    if (includeHidden) {
      _flag = 1;
    }
    if (includeInfo) {
      _flag |= 2;
    }

    internal static HiddenFilter FilterHidden = new HiddenFilter(includeHidden: false, includeInfo: true);
   
    internal static HiddenFilter FilterbOTH = new HiddenFilter(includeHidden: false, includeInfo: false);

    internal bool Contains(Severity severity) { ... }   
}

```

The simple API can be provided with a lighter weight data structure. Performance typically shows that we should avoid heavier structures like this when we know the item count is below a fixed threshold.